sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/samplers/Sampler.scala

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2023 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk
+package trace
+package samplers
+
+import org.typelevel.otel4s.trace.SpanContext
+import org.typelevel.otel4s.trace.SpanKind
+import scodec.bits.ByteVector
+
+/** A Sampler is used to make decisions on Span sampling. */
+trait Sampler {
+
+  /** Called during span creation to make a sampling result.
+    *
+    * @param parentContext
+    *   the parent's span context. `None` means there is no parent
+    *
+    * @param traceId
+    *   the trace id of the new span
+    *
+    * @param name
+    *   the name of the new span
+    *
+    * @param spanKind
+    *   the [[org.typelevel.otel4s.trace.SpanKind SpanKind]] of the new span
+    *
+    * @param attributes
+    *   the [[Attributes]] associated with the new span
+    *
+    * @param parentLinks
+    *   the list of parent links associated with the span
+    */
+  def shouldSample(
+      parentContext: Option[SpanContext],
+      traceId: ByteVector,
+      name: String,
+      spanKind: SpanKind,
+      attributes: Attributes,
+      // todo add: parentLinks: List[LinkData]
+  ): SamplingResult
```

Comment:
For the current set of implemented samplers, we only need `parentContext: Option[SpanContext]` and `traceId: ByteVector`.

However, to comply with spec requirements, we must have all these arguments.